### PR TITLE
Bug 1847996 - Migrate from kapt to KSP

### PR DESCRIPTION
**Pull request by mcarare**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1847996

### CODE REVIEW

File: android-components/plugins/dependencies/src/main/java/DependenciesPlugin.kt [hunk 1]
```diff
@@ -50,6 +50,7 @@ object Versions {
 
     const val material = "1.9.0"
     const val ksp = "1.0.11"
+    val ksp_plugin = "$kotlin-$ksp"
```
Comment (jonalmeida) [inline android-components/plugins/dependencies/src/main/java/DependenciesPlugin.kt:53]: Drive-by: This is interesting.. does this mean we are now tied to specific versions of kotlin that have ksp plugin support? What happens if we need a patch version of `kotlin` and there is no `ksp` version?

Comment (mcarare) [reply android-components/plugins/dependencies/src/main/java/DependenciesPlugin.kt:53]: That should not happen, according to the docs:

> Usually a corresponding KSP release is available within a couple of days after a new Kotlin version is released, including the [pre-releases (Beta or RC)](https://kotlinlang.org/docs/eap.html).

source: https://kotlinlang.org/docs/ksp-faq.html#how-often-do-you-update-ksp

**Review by titooan — APPROVED** [review]
Looks good to me!

File: android-components/components/lib/crash/build.gradle [hunk 2]
```diff
@@ -12,7 +12,6 @@ buildscript {
 
         dependencies {
             classpath "org.mozilla.telemetry:glean-gradle-plugin:${Versions.mozilla_glean}"
-            classpath ComponentsDependencies.plugin_ksp
```
Comment (jonalmeida) [inline android-components/components/lib/crash/build.gradle:15]: Typo or we really do not need the plugin here? I think we have room in `lib-crash` too, so it would be needed, but I don't see CI failing.

**Review by jonalmeida — APPROVED** [review]
Sorry for the slow review churn. 🙇

File: fenix/build.gradle [hunk 3]
```diff
@@ -164,7 +166,7 @@ allprojects {
         kotlinOptions.jvmTarget = "17"
         kotlinOptions.allWarningsAsErrors = true
         kotlinOptions.freeCompilerArgs += [
-            "-opt-in=kotlin.RequiresOptIn", "-Xjvm-default=all"
+            "-opt-in=kotlin.RequiresOptIn", "-Xjvm-default=all-compatibility"
```
Comment (jonalmeida) [inline fenix/build.gradle:169]: Self-note: this is fine to land because in the follow-up PR, we remove this because we use the new Room version that [has the fixes](https://android-review.googlesource.com/c/platform/frameworks/support/+/2673596).

File: android-components/components/lib/crash/build.gradle (hunk 2 above)
Comment (mcarare) [reply android-components/components/lib/crash/build.gradle:15]: This is the build script block and we do not need it for the build itself. 
We still have it below in the top-level dependencies block.
